src/clover/simulation/storage.py

```diff
@@ -620,26 +620,48 @@ def get_electric_battery_storage_profile(  # pylint: disable=too-many-locals, to
         )
 
         # Then take energy from grid if available
+        grid_energies: Dict[str, pd.DataFrame] = {}
+
         if scenario.grid:
-            grid_energy: pd.DataFrame = pd.DataFrame(
-                ((remaining_profile < 0) * remaining_profile).iloc[:, 0]  # type: ignore
-                * -1.0
-                * grid_profile.values
-            )
+            for grid_type in scenario.grid_types:
+                grid_profile = grid_profile[grid_type] # {"edl": [0, 1, 2, 3, 4]} <---- [0, 1, 2, 3, 4]
+                grid_energies[grid_type] = pd.DataFrame( # {"edl": [usage, 1, 2, 3, 4, 5]} <--- [0, 1, 2, 3, 4, 5, 6, 7]
+                    ((remaining_profile < 0) * remaining_profile).iloc[:, 0]  # type: ignore
+                    * -1.0
+                    * grid_profile.values
+                )
+
         else:
             grid_energy = pd.DataFrame([0] * (end_hour - start_hour))
+        
+        total_grid_energy= sum(grid_energies)#sum over the grid energies here
+       
         battery_storage_profile: pd.DataFrame = pd.DataFrame(
-            remaining_profile.values + grid_energy.values
+            remaining_profile.values + total_grid_energy #  Now a dictionary
         )
 
     else:
         # Take energy from grid first if available
+        # @paulharfouche: Need to do the same here.
         if scenario.grid:
-            grid_energy = pd.DataFrame(grid_profile.mul(load_energy[0]))  # type: ignore
+            for grid_type in scenario.grid_types:
+                grid_profile = grid_profile[grid_type]
+                grid_energies[grid_type] = pd.DataFrame(grid_profile.mul(load_energy[0])) 
```

Comment:
This will still be an issue right..?

If you have diesel, which is available and so the profile is `1`, and this is capable of meeting all of your demand, you don't need to draw from the EDL grid. However, if the EDL grid is also available, the code will here say "I want to meet all the demand through the diesel, and I want to meet that **same** amount of demand __again__ through the EDL grid." So, you'll want to draw from one first, not both :smile:

Reply:
I still think that, even if it's not true for Lebanon, it's best to code it to catch this error, and then set the profiles up so that there's no overlap. I'm not sure where your email comment has gone.. did you delete it?

Reply:
> Oh Okay, to clarify further, in Lebanon, EDL and Diesel do not come together. Diesel is available when EDL is 100% not available, therefore the probability of extracting from the grid becomes reliant only on the Diesel grid probability. The second EDL generates electricity, the diesel generator grid will stop and households will automatically switch to EDL (This is not decided by the households). Thus, both grids never work at the same time and have a risk of covering each others supply. Households, possess a switch called contactor that automatically switches from EDL to Diesel and vis versa, the 2 grids can't operate at the same time (because basically the diesel is here to cover when EDL is not available). For now, only at 6am and 6pm where we have a 50% probability of getting EDL, the rest is 0. However, the rest is not 100% diesel, sometimes the grid can't operate anymore and so blackout occurs (this is why we are relying on solar panels and batteries).
Having said that, what do you think would be suitable here, do we need to change it or keep it the same.
Please advise. (If anything is not clear, let me know so I explain further) 😄

Hi @paulharfouche , copying in your comment which seems to have vanished..
I think I understand what you mean. However, you need to work out how you want to code it. At the minute, the grid profiles here are availability profiles for whether the grid is available. So, there is still this issue. As the code is currently written, it will work out how much energy it can take from the grid, and then it will still go and get that same energy from the diesel if it is available... Let me know if that doesn't make sense :smile:

Reply:
Oh Okay, to clarify further, in Lebanon, EDL and Diesel do not come together. Diesel is available when EDL is 100% not available, therefore the probability of extracting from the grid becomes reliant only on the Diesel grid probability. The second EDL generates electricity, the diesel generator grid will stop and households will automatically switch to EDL (This is not decided by the households). Thus, both grids never work at the same time and have a risk of covering each others supply. Households, possess a switch called contactor that automatically switches from EDL to Diesel and vis versa, the 2 grids can't operate at the same time (because basically the diesel is here to cover when EDL is not available). For now, only at 6am and 6pm where we have a 50% probability of getting EDL, the rest is 0. However, the rest is not 100% diesel, sometimes the grid can't operate anymore and so blackout occurs (this is why we are relying on solar panels and batteries).

Having said that, what do you think would be suitable here, do we need to change it or keep it the same.
Please advise. (If anything is not clear, let me know so I explain further) 😄

Reply:
Needs further explanation

Reply:
@BenWinchester